### ISSUE TITLE
Support for seeing the download count of packages.

### ISSUE THREAD

**Issue by ShivamArora** [dart-lang/pub-dev#2714]
As a developer and owner of a package, I would like to see the download stats for my package.

The `npm` package repository on [npmjs.com](https://npmjs.com) provides the download statistics for the libraries.

I would like to see the same happening on the [pub.dev](https://pub.dev) website.

![Screenshot 2019-08-28 at 4 01 18 PM](https://user-images.githubusercontent.com/4353659/63848345-4d01ce80-c9ad-11e9-80d3-6164cc286a9f.png)

**Comment by jonasfj** [comment]
See also https://github.com/dart-lang/pub-dev/issues/1229.

------
Defining the download count is a bit non-trivial though. Should we remove bots (how), if not travis-ci will dominate the result. Do we want to measure downloads, or number of times a client asked if a newer version was available?

I think we ultimately want a more transparent popularity metric along the lines of _download count_.

**Comment by ShivamArora** [comment]
I don't think asking for a newer version should count as a download.

By download count, I meant to measure the downloads made by different projects using the `pub` tool.

However, for a reference, can we check out how npm defines their download count?

**Comment by jonasfj** [comment]
> By download count, I meant to measure the downloads made by different projects using the pub tool.

I only download a package to my machine once, then it's in my `~/.pub-cache/`, no matter how many projects I used said package in.

However, if my travis setup isn't configured to cache it can download the package for every single test run.

**Comment by ShivamArora** [comment]
> I only download a package to my machine once, then it's in my `~/.pub-cache/`, no matter how many projects I used said package in.

The same happens for npm packages as well as they are stored in `node_modules`.

It would be worth if we can check out how the npm repository is measuring the download count of packages.

**Comment by isoos** [comment]
> It would be worth if we can check out how the npm repository is measuring the download count of packages.

https://blog.npmjs.org/post/92574016600/numeric-precision-matters-how-npm-download-counts

The main gist:

> npm’s download stats are naïve by design: they are simply a count of the number of HTTP 200 responses we served that were tarball files, i.e. packages. This means the number includes:
> - automated build servers
> - downloads by mirrors
> - robots that download every package for analysis

[...]

> So the count of “downloads” is much larger than the number of people who typed “npm install yourpackage” on any given day.

[...]

> Bottom line: most packages get a trickle of downloads every day, and that’s not necessarily indicative that they’re being actively used. Only if your package is getting > 50 downloads/day can you be sure you’re seeing signal instead of noise. You will also get a burst of downloads whenever you publish a new package or a version of that package, because all the mirrors will download it.

**Comment by ShivamArora** [comment]
@isoos Do you have anything in mind regarding what could be the best approach to measure the download count?

**Comment by isoos** [comment]
> Do you have anything in mind regarding what could be the best approach to measure the download count?

@ShivamArora: I don't think there is a good approach to it, by its nature, it cannot be solved correctly. Our current popularity metric is a proxy of the total download count. A lot of thought went into it, e.g. how to filter CI systems automatically, how to balance the long-tail distribution, and I believe it is much closer to the truth than the "feel-good" download counts.

Having said that, I think it can be a valid goal to expose a few more metrics that are also a proxy for popularity, because maybe a single metric won't ever solve it:
- the "feel-good" / "raw" download counts of the packages
- the number visits of the package pages
- the number of API calls that requested data on a package (especially after these APIs are stable and public)
- search relevance / click-through-rate for a package (+ metrics on most searched-after expressions related to a package)

We are also planning to introduce some kind of community rating that also provides feedback about the quality and/or use of the package.
 
Note: we don't have any strict schedule for the above, and can't commit to it, but really open to ideas related to these.

**Comment by ShivamArora** [comment]
Cool.

I agree with the point that we should expose a few more metrics.

Having the feature of community rating would be pretty good. But don't limit it just to rating include some reviews as well.

Community Ratings and Reviews would be helpful for anyone to know about the quality of the package and what others feel about the package after using it.

Apart from that I would suggest another metric, if that's possible, which should provide information about the **Documentation Quality**.

Since good documentation is a lot more important than just using the package.

**Comment by tberghuis** [comment]
A graph how showing weekly or daily downloads is changing over time would be more useful information than absolute download counts. You could keep numbers hidden and have a graph showing relative change.

**Comment by bsutton** [comment]
@isoos 
> Our current popularity metric is a proxy of the total download count. A lot of thought went into it, e.g. how to filter CI systems automatically, how to balance the long-tail distribution, and I believe it is much closer to the truth than the "feel-good" download counts.

As a publisher I'm not certain what 'truth' you are talking about, as the current '0-100' popularity gives me no useful information.
Whilst I agree that it would be nice to filter out CI/bots etc its probably not a reality. As such a raw count actually is probably the best approach.

Raw counts will  give a sense of scale. 
The popularity score gives no sense of scale.
If I've only got 10 users then its probably time to dump the project. If there is 10K users then I'm been appreciated and its worth continuing the work :)

It is particularly unhelpful that the current metric is undocumented.
Is the range linear or exponential?
Is it somehow relative to other projects?

I have a feeling that the team tried to be too clever and we have ended up with something far worse than the raw numbers.

**Comment by mit-mit** [comment]
This is actively being considered. We're looking into options for how to get a representative count of downloads / `pub gets`, even for cases where the package is still in the local cache.

**Comment by bsutton** [comment]
@mit-mit 👍

**Comment by MaikuB** [comment]
I don't know how many authors are publishing prerelease versions but seeing download numbers for them could help authors get a sense of the number of users that helping to test out these versions. This is turn could assist with making a decision on if it's ready for a stable release

**Comment by AKushWarrior** [comment]
@mit-mit I think that when a user runs pub get, it can check the differences from the previous pubspec.lock (what packages have been added/deleted) and then push those changes to the server, which applies them to the global count. That still doesn't solve the Travis CI issue, but it does solve the cache issue.

I know you guys have looked at NPM. https://crates.io (Rust's central package repository) also provides download stats, so it might be worth checking out how they do it.

**Comment by MaikuB** [comment]
To follow up on what I had mentioned, I had changes for a plugin that was in prerelease for a number of months that was then promoted to stable having figured it had been given enough time for the community to help test and give feedback. A couple of weeks later a bug has been found. IAs this is a plugin that's used by lot of the community, at least based on what I can infer from pub.dev and GitHub, I can only assume the prereleases hadn't gotten much usage to begin with. If there is some concern around displaying download counts publicly, perhaps consider having it visible only to the authors of the plugin. Though part of what would help in this particularly is to know the amount of downloads for a particular version. An example that does this is https://www.nuget.org/

**Comment by Jjagg** [comment]
If it's feasible, I think it would be very interesting for the package author to see why their packages was installed, i.e. whether it was a direct dependency or transitive through package x.

**Comment by jayoung-lee** [comment]
+1 to more straightforward metrics. 

In the recent user study, users _used_ the popularity score and Pub Points for decision making (e.g., skip packages with <80% popularity), but they didn't necessary _understood_ the scores when using them.

**Comment by AKushWarrior** [comment]
> +1 to more straightforward metrics.
> 
> In the recent user study, users _used_ the popularity score and Pub Points for decision making (e.g., skip packages with <80% popularity), but they didn't necessary _understood_ the scores when using them.

Well, nobody understands the scores, because they're not transparent. I don't see the issue with just exposing the raw metrics: that's both more useful and more informative.

**Comment by pichillilorenzo** [comment]
Is there any news or ETA about it? Thanks!

**Comment by sigurdm** [comment]
No - we don't have an ETA for this yet - but it is something we want to get to do.

**Comment by maxxfrazer** [comment]
Any updates on when this can be added as a feature? Even if it's just an API like npm has and not shown on the package page, it's really useful to all of us to know how a package's developer base is growing (or shrinking!)

**Comment by mit-mit** [comment]
This is actively being worked on; we're adding various telemetry to enable this. But it will take a while before this is complete enough that we can start displaying it.

**Comment by gsmental** [comment]
hurry up bro,

**Comment by maxxfrazer** [comment]
> hurry up bro,

steady on, these devs don't owe you or me anything…
respectfully show your interest in this feature and move on.

**Comment by tashi146** [comment]
If the bot is what we are worried about for stats such as download count and all then at least only author should have metric option to see how package is doing. I don't think author will create bot see him self getting lot of download etc. 

Thanks dart teams

**Comment by xr0master** [comment]
> I don't think author will create bot see him self getting lot of download etc.

Why not? The downloads wrapping is a very popular cheating. The more downloads, the more often users choose your package, the more popular you become, and the more investments you get.

I wish good luck to the team to find a solution.

**Comment by alexobviously** [comment]
Any progress on this?

**Comment by ishaileshmishra** [comment]
Just checking if we have any update on the dart package statitics. 💥

**Comment by maxxfrazer** [comment]
@mit-mit is this something still on the near-term roadmap?

**Comment by incrediblezayed** [comment]
+1

**Comment by iapicca** [comment]
this feature could help sniffing "likes" from bots
- https://github.com/dart-lang/pub-dev/issues/6226

**Comment by IgorGZd** [comment]
+1

**Comment by edugemini** [comment]
+1

**Comment by pangz-lab** [comment]
+1

**Comment by ueman** [comment]
Please don't post `+1`. Use the 👍 emoji on the original post to vote it up. Comments without content don't add to the discussion and don't help prioritizing this issue. Adding a thumbs up does so.

**Comment by iapicca** [comment]
> This is actively being worked on; we're adding various telemetry to enable this. But it will take a while before this is complete enough that we can start displaying it.

@mit-mit does your [comment](https://github.com/dart-lang/pub-dev/issues/2714#issuecomment-1033661395) still reflect the status of the issue?

**Comment by mit-mit** [comment]
Yes, we're still working on it.

**Comment by renancaraujo** [comment]
Hello, 

I wanted to chime in with the insight on how game changer this would be for us at @VeryGoodOpenSource 

Two of the most popular packages that we have, [very_good_cli](https://very_good_cli.pckg.pub/) and [dart_frog_cli](https://vdart_frog_cli.pckg.pub) have thousands of reports usages (trough the now dead `usage` package) and still score relatively low in the current "popularity" metric. 


Every metric has a caveat, and I think we can live with the ones under the possible "downloads". The real impact of these caveats relies more on how you communicate it to the people reading the chart rather than making it available or not. 

As a follow-up on making "downloads" available as a metric, we can think on some capillaries that could increase the quality of life of most dart developers: 

- Identifying download spikes after a release could indicate that people are updating the package to a new version. This can also be further extended to a "version" capillarity, just like we see on[ crates.io](https://crates.io/crates/libc), see image below.
- Downloads per SDK local version could help package maintainers understand the impact of raising SDK version constraints.



Example of a download chart on [crates.io](https://crates.io/crates/libc), where the "version" capilarity is explored:


![Screenshot 2023-09-27 at 22 40 45](https://github.com/dart-lang/pub-dev/assets/6718144/c7662100-0049-46cc-9614-a401aa1694cc)

**Comment by bsutton** [comment]
I will add to @renancaraujo  comments. The version details would be particularly interesting as we could then see when a particular version of Dart no longer needs to be supported.
We currently support from dart 2.12, but how many people are still using it?
This level of knowledge would help make our development more streamlined.

**Comment by Blacktaler** [comment]
> > It would be worth if we can check out how the npm repository is measuring the download count of packages.
> 
> https://blog.npmjs.org/post/92574016600/numeric-precision-matters-how-npm-download-counts
> 
> The main gist:
> 
> > npm’s download stats are naïve by design: they are simply a count of the number of HTTP 200 responses we served that were tarball files, i.e. packages. This means the number includes:
> > 
> > * automated build servers
> > * downloads by mirrors
> > * robots that download every package for analysis
> 
> [...]
> 
> > So the count of “downloads” is much larger than the number of people who typed “npm install yourpackage” on any given day.
> 
> [...]
> 
> > Bottom line: most packages get a trickle of downloads every day, and that’s not necessarily indicative that they’re being actively used. Only if your package is getting > 50 downloads/day can you be sure you’re seeing signal instead of noise. You will also get a burst of downloads whenever you publish a new package or a version of that package, because all the mirrors will download it.

Ok, it seems a little bit harsh. But at least, are we able to see among how many users is our package popular? For example, how many users does it mean to be popular by 100%?

**Comment by mit-mit** [comment]
Assigning to @sigurdm who started investigating this

**Comment by mit-mit** [comment]
Status update: we're actively working on this, and hope to have improved metrics in at least preview by end of year. Sorry this is taking a while, but there are a lot of considerations to work through.

**Comment by dadvir** [comment]
if i can jump in, and be a bit naive and (maybe) spiritual about this, I understand the Dart comity leaders concerns regarding a well-engineered and science-like metrics.
in a real-world scenario, you've your path chosen, and most of the times, clear and simple is much more better. the current scoring tells me nothing about should i use the package. this can be risky. also can be used to exploit by black hat coders.
consider an engineer, architect or sec-dev-ops looking into a Dart package, at the moment he's completely blind regarding the usage of it.

npm metrics proved to be a great tool for the community to decide whether or not to use a package. the Dart way, 0-100 percentage, simply doesn't cut it.

i suggest to imply provide a number, and note your science like comment about scoring, and let the wise women and men choose what they will :)

**Comment by iapicca** [comment]
the PR
- https://github.com/dart-lang/pub-dev/pull/8126

seems to target this issue even if isn't mentioned directly

**Comment by bigbott** [comment]
5 years to show the **known** number of downloads. Ridiculous

**Comment by Blacktaler** [comment]
> if i can jump in, and be a bit naive and (maybe) spiritual about this, I understand the Dart comity leaders concerns regarding a well-engineered and science-like metrics. in a real-world scenario, you've your path chosen, and most of the times, clear and simple is much more better. the current scoring tells me nothing about should i use the package. this can be risky. also can be used to exploit by black hat coders. consider an engineer, architect or sec-dev-ops looking into a Dart package, at the moment he's completely blind regarding the usage of it.
> 
> npm metrics proved to be a great tool for the community to decide whether or not to use a package. the Dart way, 0-100 percentage, simply doesn't cut it.
> 
> i suggest to imply provide a number, and note your science like comment about scoring, and let the wise women and men choose what they will :)

I'm not quite sure that, what you said would be correct way too. Because there are rare packages/plugins, which are not used commonly, but necessary and not dangerous. So to work by your opinion they should add feedback section rather than likes and percentages.

**Comment by mit-mit** [comment]
> seems to target this issue even if isn't mentioned directly

Yes, we're making good progress and should have some updates soon.

**Comment by orestesgaolin** [comment]
The experimental flag seems to be available now under pub.dev/experimental

![screenshot_20241119_094800](https://github.com/user-attachments/assets/6035c676-d932-4d6f-b2d4-5a8bd88e7c52)
![screenshot_20241119_094824](https://github.com/user-attachments/assets/0b22ae2a-0b30-43e5-9b80-a63588c35bf1)
![screenshot_20241119_094930](https://github.com/user-attachments/assets/1651e27f-f00f-4a0c-8f38-9afc448913c0)

One thing that I'm not clear is for what time range the total download count is shown. Only after hovering over the number in the side bar you can see it's in the last 30 days.
![screenshot_20241119_095034](https://github.com/user-attachments/assets/0d0cf982-699b-4289-b91d-4048639d4105)

Perhaps it should also be written openly on the Scores subpage

![screenshot_20241119_095110](https://github.com/user-attachments/assets/d6b73f21-e66a-4de5-9fd0-938634411a2b)

**Comment by iapicca** [comment]
> The experimental flag seems to be available now under pub.dev/experimental

@orestesgaolin you made my day!
this could be a (or better, the final) nail in the coffin for [getx popularity](https://github.com/dart-lang/pub-dev/issues/6226#issuecomment-2486200329)

**Comment by kyle-seongwoo-jun** [comment]
Hi @mit-mit,

Great work! Thank you to the Dart/Flutter team for the amazing job. I have a quick question:

I’m curious if the download count could be included in the pub.dev API response.

```sh
$ curl https://pub.dev/api/packages/riverpod/score | jq
{
  "grantedPoints": 160,
  "maxPoints": 160,
  "likeCount": 3434,
  "popularityScore": 0.9761956555517574,
  "tags": [
    ...
  ],
  "lastUpdated": "2024-11-20T23:32:33.305499"
}
```

Currently, the response only provides points, like count, and popularity. I believe adding the download count would be a valuable enhancement.

Thank you for considering this suggestion!

Relevant Code Reference:
https://github.com/dart-lang/pub-dev/blob/dd4ecaef19388c4bf0492f1051c5cc79f3ff1ef5/app/lib/frontend/handlers/custom_api.dart#L214-L219

**Comment by sigurdm** [comment]
> Thank you for considering this suggestion!

Yeah - we will make some kind of api for fetching the counts - the details are not worked out yet though.

**Comment by vaind** [comment]
Great to see this has landed, kudos!

Any plans for version-specific downloads? Currently, when one clicks on a specific version number in the /versions page, we get the info about the package in that version (readme, example), would be great if we could, at least this way, get also downloads of a specific package version.

**Comment by sigurdm** [comment]
> Any plans for version-specific downloads?

Yes - we are planning to present those also, but this needs a bit more work.
